### ISSUE TITLE
SSD7 Variance and Steps

### ISSUE THREAD

**Issue by cyberphantom** [pierluigiferrari/ssd_keras#148]
Hi, I really liked your model and it train very fast on my dataset and without overfitting,.
I have some questions about your settings

1. I noticed that you changed the variances values all to 1s. Even that I didn't understand well why we need to use them and what the appropriate mechanize to assign them from the first place, So how did you picked up these values?

2. Here you are using only 1 aspect ration per layer, is this will make the bounding boxes comparison easier and faster or is there any other purpose in assigning that?

3. What is the benefit if we use the steps again per each classification layer in SS7 instead of assigning steps to None?

4. Your data augmentation enhanced the results a lot for SSD7, what about if we trying them on SSD300 for example?

5. I tried to use SSD7 with image height = 360 and width= 640 but it feels for some reason, so I went back to 300x 300 resize option 

6. What about using dense blocks and I mean by that passing layers from previous conv to the next one until the last classifier in SSD7?

Thanks,

**Comment by pierluigiferrari** [comment]
1. Setting all variances to 1 effectively takes them out of the equation (or assumes unit variance for the anchor offset distributions, depending on how you want to interpret them). I made this choice rather arbitrarily when I built SSD7 simply because the necessity of assuming variances smaller than 1 wasn't clear to me. To the best of my knowledge, Wei Liu et al. don't provide any real justification for the variances they used for the original models and I haven't spent a lot of time trying to better understand this parameter and its effect. Since 1 seemed to work well enough, I just left it at that. It's possible that other variance values may lead to improved learning, I haven't tried. As the SSD7 training notebook says, SSD7 is not a carefully tuned model, I just needed it to be quickly trainable to use it for debugging.

2. SSD7 as preset in the training notebook uses four aspect ratios in every predictor layer: one box with aspect ratio 0.5, two with 1.0, and one 2.0. Fewer aspect ratios means fewer anchor boxes, means a less complex model, means faster computation and smaller memory footprint. More aspect ratios might mean (depending on the dataset) better fit between anchor boxes and ground truth object shapes, might mean improved learning. Roughly. So it's a trade-off that depends on the dataset you're training on.

3. If you set the steps to `None`, then the anchor boxes will be distributed canonically across the spatial grid. This means that if a given predictor layer has spatial dimensions `(h, w)` and the image has size `(img_h, img_w)`, then the anchor boxes for that layer will be centered at every `img_h / h` pixels in the vertical dimension and at every `img_w / w` pixels in the horizontal direction, starting at `img_h / h / 2` and `img_w / w / 2` pixels, respectively. Note that the resulting values need not be integers. This should theoretically be a better spacing for the anchor boxes than the fixed step size that the original Caffe implementation uses, because the canonical spacing explained above should align the anchor boxes more consistently with the receptive fields of the neurons, but I haven't done any proper comparisons.

4. The data augmentation pipeline that is preset in the SSD7 training notebook should actually be less good than the one described in the SSD paper, because it doesn't contain random deformations. The reason why I was using it is because it is a lot faster than the original SSD data augmentation pipeline.

5. This point doesn't contain a question. Also, I don't understand it.

6. Are you talking about dense blocks as in DenseNet? Sure, that would probably be a good idea. It's important to understand that Wei Liu et al. created SSD a while ago and at that point, VGG might have been one of the best base network architectures for their purpose, but obviously there exist much better architectures today, so there is no reason not to try to put the SSD architecture on top of more advanced CNN architectures.

**Comment by cyberphantom** [comment]
Thanks for your detailed clarification on my concerns. Probably I'm gonna try DenseNet and if there is a luck in performance, I will let you know.
In point 5, sorry for my typo, what I mean is that as you are using images (480x300) pixel  for height and width. I tried my own data with their actual size (640x360) but the system fails after reaching to the 50s steps (out of 1000) in the first epoch.
Then, when I resized them in the creation of the hdf5 files to (300x300) it works well. So why you didn't resize your data and prefered to use (300x480) height and width image size.

**Comment by cyberphantom** [comment]
Another thing that I want to ask you is how can I cite your work? I couldn't find any publication regarding SSD7.

**Comment by mattroos** [comment]
@pierluigiferrari, can you point to the location in the Wei Liu et al. paper they discuss the variance and their chosen values? I can't find anything. Do they use different terminology? Is it only in the code rather than the paper? Also, in one of your responses to a different but related issues you referred to a distribution. Is it really a variance in a probabilistic sense? I'm struggling to see the difference between using the variance as a divisor of the box coordinates versus just using different box scales and/or aspect ratios. I'm clearly missing something fundamental.

**Comment by mattroos** [comment]
@pierluigiferrari and everyone else, I found ths recent [blog post](https://leimao.github.io/blog/Bounding-Box-Encoding-Decoding/) (April 2019) by Lei Mao that gives a detailed explanation of "variance" in SSD implementations. Very helpful, and it even links to this repo as one point. I initially found the blog link in an [issue discussion at this repo](https://github.com/rykov8/ssd_keras/issues/53), which had some additional helpful comments.

**Comment by pierluigiferrari** [comment]
@mattroos thanks for this blog post! One thing to note is that I've implemented the variances exactly as in the original C++/Caffe implementation by Wei Liu. As you've probably found out in the meantime, the variances only show up in the code, they are not mentioned in the paper - a glaring omission for sure. If you interpret the anchor/prior box coordinates as the expected values of the ground truth coordinates, then the offset of each coordinate is indeed a random variable with expected value zero and some standard deviation and dividing by this standard deviation normalizes the distribution. I would speculate that whoever first had the idea to introduce this parameter in bounding box regression hasn't been thinking too much about whether this is a standard deviation or variance, it's just that the name "variance" is easier to use as a parameter name and is sufficiently descriptive. I may be wrong. The point in the blog post about it being the standard deviation rather than the variance may be theoretically correct, but is a moot point practically: The mathematics doesn't care what you call it, what matters is that the numbers chosen are good proxies. In that sense, [0.1, 0.1, 0.2, 0.2] either are good values or they aren't, regardless of them being named "variances". We can only assume that the authors systematically found these values to be optimal or at least close to optimal.

**Comment by mattroos** [comment]
@pierluigiferrari, my current understanding is exactly as you describe. Good to get your confirmation! I agree about the semantics of "standard deviation" versus "variance." Though it could have some slight impact if the original authors computed the empirical variances to be [0.1, 0.1, 0.2, 0.2] but then treated them as empirical standard deviations. That is, if we assume that encoded targets with standard deviation of 1.0 result in improved learning (faster convergence rate and/or lower loss) versus some other standard deviation. But it probably makes very little difference in the end even if that assumption is true.

Thanks!
